File: typescript/nodes/countNode.test.ts
import { assertEquals } from "@std/assert";
import * as CountNode from "./countNode.ts";

Deno.test("countNode should increment count and transition to onCount when random <= 0.7", async () => {
  const state = { count: 5 };
  const utils: CountNode.Utils = {
    log: () => {},
    sleep: () => Promise.resolve(),
    random: () => 0.5, // 0.5 <= 0.7 is true, so stop = false
  };
  const node = CountNode.create({ onCount: "count", onStop: "stop" }, utils);
  const [next, nextState] = await node(state, "count");
  assertEquals(next, "count");
  assertEquals(nextState.count, 6);
});

Deno.test("countNode should increment count and transition to onStop when random > 0.7", async () => {
  const state = { count: 5 };
  const utils: CountNode.Utils = {
    log: () => {},
    sleep: () => Promise.resolve(),
    random: () => 0.8, // 0.8 > 0.7 is true, so stop = true
  };
  const node = CountNode.create({ onCount: "count", onStop: "stop" }, utils);
  const [next, nextState] = await node(state, "count");
  assertEquals(next, "stop");
  assertEquals(nextState.count, 6);
});

File: go/nodes/count_node_test.go
package nodes

import (
	"context"
	"testing"
	"time"

	"github.com/argenkiwi/ambler/go"
)

type mockCountNodeUtils struct {
	randomVal float64
}

func (m mockCountNodeUtils) Log(message string) {}

func (m mockCountNodeUtils) Sleep(ctx context.Context, duration time.Duration) error {
	return nil
}

func (m mockCountNodeUtils) Random() float64 {
	return m.randomVal
}

func TestCountNode(t *testing.T) {
	ctx := context.Background()
	edges := CountNodeEdges[string]{
		OnCount: ambler.Ptr("count"),
		OnStop:  ambler.Ptr("stop"),
	}

	t.Run("increments count and continues when random <= 0.7", func(t *testing.T) {
		utils := mockCountNodeUtils{randomVal: 0.5}
		node := CreateCountNode[mockState, string](edges, utils)
		state := mockState{count: 5}

		next, err := node(ctx, state, "count")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}

		if next.NodeID == nil || *next.NodeID != "count" {
			t.Errorf("expected count, got %v", next.NodeID)
		}
		if next.State.GetCount() != 6 {
			t.Errorf("expected 6, got %d", next.State.GetCount())
		}
	})

	t.Run("increments count and stops when random > 0.7", func(t *testing.T) {
		utils := mockCountNodeUtils{randomVal: 0.8}
		node := CreateCountNode[mockState, string](edges, utils)
		state := mockState{count: 5}

		next, err := node(ctx, state, "count")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}

		if next.NodeID == nil || *next.NodeID != "stop" {
			t.Errorf("expected stop, got %v", next.NodeID)
		}
		if next.State.GetCount() != 6 {
			t.Errorf("expected 6, got %d", next.State.GetCount())
		}
	})
}
